File: washee_box/exception_handling/exception_handler.py
import os
import pytz
import inspect
from pathlib import Path
from datetime import datetime

class ExceptionHandler:
    # ExceptionHandler is a class where all try/exceåt blocks in the codebase use the .handle() method on any catched Exception.
    # This is to reuse and standardise how errors are handled across the codebase.
    
    def __init__(self):
        self.debug = False if int(os.environ.get("BOX_DEBUG", default="1")) == 0 else True
        self.log_location = os.path.join(Path(__file__).parent, "exception_log.txt")
        with open(self.log_location, "a") as f:
            f.write("ExceptionHandler initialized at: " + datetime.now(pytz.timezone('Europe/Copenhagen')).strftime("%Y/%m/%d, %H:%M:%S") + "\n")
    
    def reset_log(self, forced=False):
        """Resetting the log should only be possible in debug mode or if one is very sure in production"""
        if self.debug or forced:
            with open(self.log_location, "w") as f:
                f.write("ExceptionHandler reset log file at: " + datetime.now(pytz.timezone('Europe/Copenhagen')).strftime("%Y/%m/%d, %H:%M:%S") + "\n")
        else:
            with open(self.log_location, "a") as f:
                f.write("Failed to reset the log. Can only be reset in debug mode or if using the forced parameter. reset_log was called by this function: " + str(inspect.stack()[1][3]) + "\n")
    
    def handle(self, exception, log=False,show=False,crash=False):
        if not log and not show and not crash:
            raise Exception("ExceptionHandler not configured properly.\n" +
                        "ExceptionHandler has to have set at least one parameter of (log, show or crash) to True\n" +
                        "Error occured with the following exception:\n" +
                        str(exception))
            
        if log:
            with open(self.log_location, "a") as f:
                f.write(str(exception))
        if show:
            print(str(exception))
        if crash:
            raise Exception(exception)

File: washee_app/lib/core/errors/exception_handler.dart
import 'dart:io';

import 'package:washee/core/helpers/web_communicator.dart';
import 'package:washee/core/network/network_info.dart';
import 'package:washee/injection_container.dart';
import 'package:washee/core/helpers/date_helper.dart';

class ExceptionHandler{
  // ExceptionHandler is a class where all try/catch blocks in the codebase use the .handle() method on any catched Exception.
  // This is to reuse and standardise how errors are handled across the codebase.

  // Flutter will highly likely have its own way of handling exceptions.
  // It would be nice though, if these ways of handling exceptions are "wrapped" in a custom class
  // So everyone knows where to look for, and improve, security "stuff", rather than it being spread out
  String logLocation = Directory.current.path + "/lib/core/errors/exception_log.txt";

  ExceptionHandler();

  void resetLog(){
    File(logLocation).writeAsStringSync("ExceptionHandler reset log file at: " + DateHelper().currentTime().toString() + "\n", mode: FileMode.write);
  }

  String handle(String exception, {log=false, show=false, crash=false}) {
    //File(logLocation).writeAsStringSync("ExceptionHandler initialized at: " + DateHelper.currentTime().toString() + "\n", mode: FileMode.append);

    if (log==false && show==false && crash==false){
      throw new Exception("ExceptionHandler not configured properly.\n" +
                        "ExceptionHandler has to have set at least one parameter of (log, show or crash) to True\n" +
                        "Error occured with the following exception:\n" +
                        exception.toString());
    }

    if (log){
      try{
        sl<NetworkInfo>().isConnected.then((connected) => {
          if(connected){
            sl<WebCommunicator>().postLog(exception)
          }
      });
      }
      catch (e){
        print("Error could not be send to logs: " + exception);
      }
    }
    if (show){
      print(exception.toString());
    }
    if (crash){
      throw new Exception(exception.toString());
    }

    return exception.toString();
  }
}